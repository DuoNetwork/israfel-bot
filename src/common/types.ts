export * from '../../../duo-admin/src/common/types';
export * from '../../../duo-contract-wrapper/src/types';
export {
	IWsResponse,
	IWsOrderBookResponse,
	IWsOrderResponse,
	IWsRequest,
	IWsInfoResponse,
	IOrderBookSnapshot,
	IToken,
	IWsAddOrderRequest,
	IOrderBookSnapshotUpdate,
	IWsOrderBookUpdateResponse,
	IOrderBookSnapshotLevel,
	IWsOrderHistoryRequest,
	IStringSignedOrder
} from '../../../israfel-relayer/src/common/types';

export interface IOption {
	live: boolean;
	token: string;
	baseToken: string;
	type: string;
	tenor: string;
	source: string;
	provider: string;
	debug: boolean;
	env: string;
}

export interface IBaseMarketData {
	source: string;
	base: string;
	quote: string;
	timestamp: number;
}

export interface IPrice extends IBaseMarketData {
	period: number;
	open: number;
	high: number;
	low: number;
	close: number;
	volume: number;
}

export interface IAccounts {
	address: string;
	privateKey: string;
}
