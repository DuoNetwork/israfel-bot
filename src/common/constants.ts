export * from '../../../duo-admin/src/common/constants';
export {
	WS_UNSUB,
	DB_ORDERS,
	WS_INFO,
	DB_ORDER_BOOKS,
	WS_OK,
	DB_SNAPSHOT,
	WS_SUB,
	RELAYER_ADDR_KOVAN,
	RELAYER_ADDR_MAIN,
	DB_ADD,
	TOKEN_WETH,
	ONE_MINUTE_MS
} from '../../../israfel-relayer/src/common/constants';

export const LOG_INFO = 'INFO';
export const LOG_DEBUG = 'DEBUG';
export const LOG_ERROR = 'ERROR';
export const LOG_RANKING: { [level: string]: number } = {
	[LOG_ERROR]: 0,
	[LOG_INFO]: 1,
	[LOG_DEBUG]: 2
};

export const SRC_INFURA = 'infura';
export const PROVIDER_INFURA_MAIN = 'https://mainnet.infura.io';
export const PROVIDER_INFURA_KOVAN = 'https://kovan.infura.io';
export const MAKE_DEPTH = 'makeDepth';

export const API_BITFINEX = 'bitfinex';
export const API_GEMINI = 'gemini';
export const API_KRAKEN = 'kraken';
export const API_GDAX = 'gdax';

export const SRC = [
	// API_BITFINEX,
	API_GEMINI,
	API_KRAKEN,
	API_GDAX
];

export const MIN_ETH_BALANCE = 0.5;
export const MIN_WETH_BALANCE = 0.4;
export const MIN_TOKEN_BALANCE = 20;
export const MAX_TOKEN_BALANCE = 100;

export const PRICE_STEP = 0.001;
export const PRICE_LEVEL = 5;

export const DB_PRICES_PRIMARY_KEY_RESOLUTION: {
	[period: number]: 'minute' | 'hour' | 'day' | 'month';
} = {
	0: 'minute',
	1: 'hour',
	10: 'hour',
	60: 'day',
	360: 'month',
	1440: 'month'
};
