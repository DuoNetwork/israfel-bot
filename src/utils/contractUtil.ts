import moment from 'moment';
import Web3 from 'web3';
import DualClassWrapper from '../../../duo-contract-wrapper/src/DualClassWrapper';
import Web3Wrapper from '../../../duo-contract-wrapper/src/Web3Wrapper';
import Web3Util from '../../../israfel-relayer/src/utils/Web3Util';
import * as CST from '../common/constants';
import { IAccounts, IDualClassStates, IOption } from '../common/types';
import priceUtil from './priceUtil';
import util from './util';

const Tx = require('ethereumjs-tx');

export class ContractUtil {
	public dualClassCustodianWrapper: DualClassWrapper;
	public web3Wrapper: Web3Wrapper | null = null;
	public web3Util: Web3Util;
	public web3: Web3;

	constructor(web3Util: Web3Util, web3Wrapper: Web3Wrapper, option: IOption) {
		this.web3Wrapper = web3Wrapper;
		this.web3Util = web3Util;
		this.dualClassCustodianWrapper = new DualClassWrapper(
			web3Wrapper,
			web3Wrapper.contractAddresses.Custodians.Beethoven[option.tenor].custodian.address
		);
		this.web3 = new Web3(
			option.source
				? new Web3.providers.HttpProvider(option.provider)
				: new Web3.providers.WebsocketProvider(option.provider)
		);
	}

	public async estimateBeethovenPrice(option: IOption): Promise<number[]> {
		if (!this.dualClassCustodianWrapper) {
			util.logDebug(`no dualClassWrapper initiated`);
			return [];
		}

		const states: IDualClassStates = await this.dualClassCustodianWrapper.getStates();

		const price = await priceUtil.getLastPrice(15, option);

		if (!price) {
			util.logDebug(`no price calculated`);
			return [];
		}

		const resetTime = Math.floor(states.resetPriceTime / 1000);
		const { resetPrice, beta, alpha, periodCoupon } = states;

		const time = Math.floor(moment.utc().valueOf() / 1000);

		const numOfPeriods = Math.floor((time - resetTime) / (states.period / 1000));
		const navParent = (price / resetPrice / beta) * (1 + alpha);

		const navA = 1 + numOfPeriods * Number(periodCoupon);
		const navAAdj = navA * alpha;
		if (navParent <= navAAdj) return [navParent / alpha, 0];
		else return [navA / price, navParent - navAAdj / price];
	}

	public async estimateBeethovenTokenCreateAmt(ethAmount: number): Promise<number[]> {
		if (!this.dualClassCustodianWrapper || ethAmount <= 0) {
			util.logDebug(`no dualClassWrapper initiated`);
			return [];
		}
		const states: IDualClassStates = await this.dualClassCustodianWrapper.getStates();
		const tokenValueB =
			(((1 - states.createCommRate) * states.resetPrice) / (1 + states.alpha)) * ethAmount;
		const tokenValueA = states.alpha * tokenValueB;
		return [tokenValueA, tokenValueB];
	}

	private getMainAccount(): IAccounts {
		const faucetAccount = require('../keys/faucetAccount.json');

		return {
			address: faucetAccount.publicKey,
			privateKey: faucetAccount.privateKey
		};
	}

	private async ethTransferRaw(
		web3: Web3,
		from: string,
		privatekey: string,
		to: string,
		amt: number,
		nonce: number
	) {
		const rawTx = {
			nonce: nonce,
			gasPrice: web3.utils.toHex((await web3.eth.getGasPrice()) || CST.DEFAULT_GAS_PRICE),
			gasLimit: web3.utils.toHex(23000),
			from: from,
			to: to,
			value: web3.utils.toHex(web3.utils.toWei(amt.toPrecision(3) + '', 'ether'))
		};
		return web3.eth
			.sendSignedTransaction('0x' + this.signTx(rawTx, privatekey))
			.then(receipt => util.logInfo(JSON.stringify(receipt, null, 4)));
	}

	public signTx(rawTx: object, privateKey: string): string {
		try {
			const tx = new Tx(rawTx);
			tx.sign(new Buffer(privateKey, 'hex'));
			return tx.serialize().toString('hex');
		} catch (err) {
			util.logError(err);
			return '';
		}
	}

	public async checkBalance(
		pair: string,
		tokenIndex: number,
		addresses: string[]
	): Promise<string[]> {
		const [code1, code2] = pair.split('|');

		for (const address of addresses)
			if (this.web3Util) {
				// ethBalance
				const ethBalance = await this.web3Util.getEthBalance(address);
				util.logInfo(`the ethBalance of ${address} is ${ethBalance}`);
				if (ethBalance < CST.MIN_ETH_BALANCE) {
					util.logDebug(
						`the address ${address} current eth balance is ${ethBalance}, make transfer...`
					);
					addresses = addresses.filter(addr => addr !== address);
					const faucetAccount: IAccounts = this.getMainAccount();

					await this.ethTransferRaw(
						this.web3,
						faucetAccount.address,
						faucetAccount.privateKey,
						address,
						CST.MIN_ETH_BALANCE,
						await this.web3Util.getTransactionCount(faucetAccount.address)
					);
					continue;
				}

				// wEthBalance
				const wEthBalance = await this.web3Util.getTokenBalance(code2, address);
				if (wEthBalance < CST.MIN_WETH_BALANCE) {
					util.logInfo(
						`the address ${address} current weth balance is ${wEthBalance}, wrapping...`
					);
					const amtToWrap = CST.MIN_WETH_BALANCE + 0.1 - wEthBalance;
					if (ethBalance > amtToWrap) {
						util.logInfo(`start wrapping for ${address} with amt ${amtToWrap}`);
						await this.web3Util.wrapEther(amtToWrap, address);
					} else {
						util.logInfo(
							`the address ${address} current weth balance is ${wEthBalance}, not enought eth to wrapping, skip...`
						);
						addresses = addresses.filter(addr => addr !== address);
						continue;
					}
				}

				// tokenBalance
				const tokenBalance = await this.web3Util.getTokenBalance(code1, address);
				util.logInfo(`the ${code1} tokenBalance of ${address} is ${tokenBalance}`);
				const accountsBot: IAccounts[] = require('../keys/accountsBot.json');
				const account = accountsBot.find(a => a.address === address);
				const gasPrice = Math.max(
					await this.web3Util.getGasPrice(),
					CST.DEFAULT_GAS_PRICE * Math.pow(10, 9)
				);
				if (tokenBalance < CST.MIN_TOKEN_BALANCE) {
					util.logInfo(
						`the address ${address} current token balance of ${code1} is ${tokenBalance}, need create more tokens...`
					);

					const tokenValues = await this.estimateBeethovenTokenCreateAmt(
						ethBalance - CST.MIN_ETH_BALANCE - 0.1
					);
					if (
						tokenValues.length &&
						tokenValues[tokenIndex] + tokenBalance > CST.MIN_TOKEN_BALANCE
					) {
						util.logDebug(`creating token ${code1}`);
						if (account)
							await this.dualClassCustodianWrapper.createRaw(
								address,
								account.privateKey,
								gasPrice,
								CST.CREATE_GAS,
								Number(
									Number(
										(ethBalance - CST.MIN_ETH_BALANCE - 0.1).toFixed(10)
									).toPrecision(3)
								)
							);
						else {
							util.logDebug(`the address ${address} cannot create, skip...`);
							addresses = addresses.filter(addr => addr !== address);
							continue;
						}
					}
				} else if (tokenBalance >= CST.MAX_TOKEN_BALANCE) {
					util.logDebug(
						`the address ${address} current token balance of ${code1} is ${tokenBalance}, need redeem back...`
					);
					if (account) {
						const states: IDualClassStates = await this.dualClassCustodianWrapper.getStates();
						await this.dualClassCustodianWrapper.redeemRaw(
							address,
							account.privateKey,
							tokenBalance - CST.MAX_TOKEN_BALANCE,
							(tokenBalance - CST.MAX_TOKEN_BALANCE) / states.alpha,
							gasPrice,
							CST.REDEEM_GAS
						);
					}
				}

				// tokenAllowance
				const tokenAllowance = await this.web3Util.getProxyTokenAllowance(code1, address);
				if (tokenAllowance <= 0) {
					util.logDebug(`the address ${address} token allowance is 0, approvaing.....`);
					this.web3Util.setUnlimitedTokenAllowance(code1, address);
				}
			}

		return addresses;
	}
}
