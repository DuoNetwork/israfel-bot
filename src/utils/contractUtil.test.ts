// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import Web3Wrapper from '../../../duo-contract-wrapper/src/Web3Wrapper';
import Web3Util from '../../../israfel-relayer/src/utils/Web3Util';
import * as CST from '../common/constants';
import { ContractUtil } from './contractUtil';
const mnemonic = require('../keys/mnemomic.json');

const web3Util = new Web3Util(null, false, mnemonic.mnemomic, false);
const option = {
	live: false,
	source: 'infura',
	provider: 'provider',
	token: 'token',
	type: 'Beethoven',
	tenor: 'Perpetual'
} as any;
const web3Wrapper = new Web3Wrapper(null, option.source, option.provider, option.live);
const contractUtil = new ContractUtil(web3Util, web3Wrapper, option);

test('checkBalance, not enought ethBalance', async () => {
	contractUtil.web3Util.getEthBalance = jest.fn(() => CST.MIN_ETH_BALANCE - 0.2);

	contractUtil.web3Util.getTokenBalance = jest.fn(() => 200);
	contractUtil.web3Util.getProxyTokenAllowance = jest.fn(() => 100000);
	contractUtil.web3Util.getGasPrice = jest.fn(() => 1000000000);

	contractUtil.web3Util.getTransactionCount = jest.fn(() => 10);
	contractUtil.web3Wrapper.ethTransferRaw = jest.fn(() => Promise.resolve());
	await contractUtil.checkBalance('pair', 0, ['address0']);
	expect((contractUtil.web3Wrapper.ethTransferRaw as jest.Mock).mock.calls).toMatchSnapshot();
});
