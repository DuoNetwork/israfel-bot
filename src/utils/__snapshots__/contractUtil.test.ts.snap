// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkBalance, not enought ethBalance 1`] = `
Array [
  Array [
    "0x00D8d0660b243452fC2f996A892D3083A903576F",
    "16ea9a15dd4381a2ea9c41531fee530e6a78dadc1a86d3e64f207c35f4be1d17",
    "address0",
    3,
    10,
  ],
]
`;
